src/main/java/org/candlepin/model/EntitlementCurator.java

```diff
@@ -562,6 +562,51 @@ private int countMatchesByFilters(AbstractHibernateObject<?> object, String obje
         return this.cpQueryFactory.buildQuery(this.currentSession(), criteria);
     }
 
+    /**
+     * Fetches a the entitlements used by consumers in the specified environment.
+     *
+     * @param environmentId
+     *  The ID of the environment for which to fetch entitlements
+     *
+     *  @param contentIds
+     *  A collection of content Ids for which to fetch entitlements
+     *
+     * @return List of entitlement ids
+     */
+    public List<String> listEntitlementIdByEnvironmentAndContent(String environmentId,
+        Collection<String> contentIds) {
+        String sql =
+            "select a.id from cp_entitlement a " +
+            "    join cp_consumer b on a.consumer_id=b.id " +
+            "    join cp_environment c on b.environment_id=c.id " +
+            "    join cp_pool d on a.pool_id=d.id " +
+            "    join cp2_product_content f on d.product_uuid=f.product_uuid " +
+            "    join cp2_content g on f.content_uuid=g.uuid " +
+            "where c.id=:environmentId"  +
+            "    and g.content_id in (:contentIds) ";
+        List<String> results = this.getEntityManager()
+            .createNativeQuery(sql)
+            .setParameter("environmentId", environmentId)
+            .setParameter("contentIds", contentIds)
+            .getResultList();
+
+        sql =
+            "select a.id from cp_entitlement a " +
+            "    join cp_consumer b on a.consumer_id=b.id " +
+            "    join cp_environment c on b.environment_id=c.id " +
+            "    join cp2_pool_provided_products f on a.pool_id=f.pool_id " +
```

Comment:
This table `cp2_pool_provided_products` is deprecated. The only reason it still exists is for backup, until Hosted migrates to 4.0 and we validate 4.0 works, at which point we will delete it. 

You have to use the new table `cp2_product_provided_products` (since now the provided products have marketing products be their parents, not pools). You likely need 1 more join here, but the way you get from Entitlement to Content now should be: cp_entitlement -> cp_pool -> cp2_product_provided_products -> cp2_product_content -> cp2_content

Reply:
Also, I was wondering why spec tests don't fail for this (specificaly `regenerates certificates after promoting/demoting content` in `environment_cert_v3_spec.rb`), and it turns out it's because it is currently assigning content directly to the marketing product (which means it's the first query in this method that will return the results), instead of creating an engineering product to add to the marketing product, and then adding the content to the engineering product (which is the realistic scenario); so let's fix that test too